### PR TITLE
[FLINK-19748] Iterating key groups in raw keyed stream on restore fails if some key groups weren't written

### PR DESCRIPTION
**Pull request by tzulitai**
This PR is based on top of #13761, since a new test was added in #13761 that would also cover the fix in this PR.
Only the last 2 commits are relevant.

## What is the purpose of the change

When iterating through key groups in restored raw keyed state streams, the iterator (`StreamTaskStateInitializerImpl#KeyGroupStreamIterator`) doesn't respect that fact that some key groups were skipped and were not written, .e.g:
```
// on snapshots
KeyedStateCheckpointOutputStream rawKeyedStream = snapshotContext.getRawKeyedOperatorStateOutput();
rawKeyedStream.startNewKeyGroup(0);
rawKeyedStream.startNewKeyGroup(3); // skipped key groups 1 and 2
```

In this case, the offsets for key groups 1 and 2 written in the checkpoint metadata will be `-1` (meaning, an undefined offset).
On restore, however, the iterator still attempts to seek the restored keyed state streams with the offsets, ultimately failing with a position out of bounds when it attempts to seek to position `-1`.

The usage behaviour of skipping key groups when writing to the raw keyed state stream is currently completely valid as the `KeyedStateCheckpointOutputStream` API does not forbid it. StateFun, for example, did exactly that.

## Solution

This PR solves the issue by modifying `StreamTaskStateInitializerImpl#KeyGroupStreamIterator` to skip key groups with undefined offsets (i.e. `-1`).
The changes are isolated such that is should only affect the restore of raw keyed state streams, and does not affect the restore of state backends.

## Brief change log

- a130d84 Wraps the original key group offsets iterator to filter out key groups with undefined offsets.
- c6bae68 Adjusts the existing `AbstractStreamOperatorTest.testCustomRawKeyedStateSnapshotAndRestore` test to only write some key groups, and not all of them.

## Verifying this change

- The new `AbstractStreamOperatorTest.testCustomRawKeyedStateSnapshotAndRestore` test should pass.
- There are existing integration tests that test snapshotting and restoring of heap-based timers, which should cover the current usage of the raw keyed state stream in Flink.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorTest.java [hunk 1]
```diff
@@ -444,10 +445,11 @@ public void testCustomRawKeyedStateSnapshotAndRestore() throws Exception {
 		final int maxParallelism = 10;
 		final int numSubtasks = 1;
 		final int subtaskIndex = 0;
-		final KeyGroupRange keyGroupRange = KeyGroupRange.of(0, maxParallelism - 1);
+		final List<Integer> keyGroupsToWrite = Arrays.asList(2, 3, 8);
```
Comment (carp84) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorTest.java]: Maybe we could extract the common logic out as a utility test method and add another test case in name of like `testCustomRawKeyedStateSnapshotAndRestoreWithPartialWrittenKeyGroups`?

**Review by carp84 — CHANGES_REQUESTED** [review]
Thanks for the efforts @tzulitai ! The changes overall LGTM, just a minor comment about the test case update.

And I'd suggest we rebase the PR after #13761 is in, to better reflect what has been modified here. :-)

**Comment by tzulitai** [comment]
Thanks for the review @carp84!
I've addressed your comments, and will rebase this after #13761 is merged.

**Comment by tzulitai** [comment]
@flinkbot run azure

**Comment by tzulitai** [comment]
@flinkbot run azure

**Review by carp84 — APPROVED** [review]
+1, the latest changes LGTM.

It seems #13761 needs some more time, either waiting for it and rebase here or committing this one first (then we need a rebase there) works for me.

**Pull request closed without merging**